keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+namespace from `NamespaceTemplate`, it has no idea who created the namespace.
+To be able to substitute `$(CREATOR)`, a `MutatingAdmissionWebhook` is involved in namespace
+creation request. 
+It adds an annotation `policy/namespace-creator=name` to the namespace being created.
+This annotation is immutable from further update/patch requests.
+
+#### Readiness of a namespace
+
+A newly created namespace is not ready before all objects in `NamespaceTemplate`
+are fully populated.
+To determine the readiness of a namespace, simply inspect `phase` in `status`
+subresource of the namespace. It's ready when the `phase` is `Active`.
+Alternatively, the `NamespaceTemplate` controller is registered as an _Initialization Controller_
+and uses `Initializer` admission control to hide the namespace before it's fully populated.
+
+### Security Consideration
+
+#### Opt-out is privileged
+
+A user who has the permission to create a namespace can put the opt-out annotation,
+though with proper RBAC setup, 
+the user may not have further permission in that namespace without being populated.
+The situation can be handled in a better way with a `ValidatingAdmissionWebhook` checking
+a specific RBAC binding:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: namespace-template-optout
+rules:
+  - apiGroups: ["policy"]
+    resources: ["namespacetemplates"]
+    verbs: ["optout"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: namespace-template-optout
+roleRef:
+  kind: ClusterRole
+  name: namespace-template-optout
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "permitted-user"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+The webhook will fail the request if the opt-out annotation is present without the RBAC binding for the requestor.
+
+#### Race Condition
+When a namespace is created, there's a gap before all objects in the namespace
+templates are populated.
+During the gap, the users may have the access to create objects into namespaces.
+
+To avoid the situation, a few mechanisms can be used:
+
+##### Creating namespace through service
+
+Users don't have permission to create namespace directly using Kubernetes API.
+The namespace creation is performed using a custom service which runs under a
+service account with enough privilege to create namespaces and objects inside namespaces.
+The service will also grant the current user privilege to manage the namespace.
+
+This approach requires additional development effort and introduces non-standard way
+for namespace creation.
+
+##### Setup RBAC properly
+
+See above [Self-serviced namespace creation](#self-serviced-namespace-creation)
+for details of RBAC setup.
+
+This approach requires pre-setup work from cluster admin for ClusterRoleBinding to grant
+users/groups the privilege to create namespaces.
+Beyond that, it is the simplest and recommended approach.
+
+##### Use of Initializers
```

Comment:
I feel like just creating namespace through delegated service avoids the need for this complexity

Reply:
With help of a delegated service, it makes the problem simpler. However that requires the delegated service, with concerns in some use cases:

1. The user experience is inconsistent using delegated service, as in different environment/teams/corps, the delegated services are different
2. The cluster admin will either spend effort developing their own delegated service or use an existing solution, and they have to educate the users of the cluster about how to create a namespace.

In a self-service namespace creation use case, the users of the cluster want to simply use "kubectl" to create namespaces.